### ISSUE TITLE
Advanced DataGrid

### ISSUE THREAD

**Issue by alisonailea** [Esri/calcite-components#6646]
### Description

Differentiating from a Simple DataGrid, this component should provide advanced patterns for data grid interaction and functionality. This component may wrap or extend a 3rd party library to get the required acceptance criteria to production faster with less maintenance costs. Additional research is still required. 

### User Stories

As a consumer of Calcite I want to 
- virtualize creating a DataGrid from a large data set.
- be able to easily render any component inside a DataGrid component.
- have a fast and performant way to find data in existing loaded data set. 
- be able to select table rows.
- be able to select table columns
- have a built-in loading state for initialization of large data sets
- have lazy-loading on scroll of additional data rows
- have lazy-loading on scroll of additional data columns
- be able to expand a row to show additional information about the row-set
- wrap content inside a cell
- update / replace / add data to a DataGrid asyncronysly

### Acceptance Criteria

The component should Render large data sets
The component should provide a calcite-search to help find data in the set
The component should provide a method for updating the data set
The component should provide a method for replacing the data set
The component should provide a method for adding to the data set
The component should provide a UI for a loading state at initialization. 
The component should allow for lazy-loading additional columns on scroll.
The component should allow for lazy-loading additional rows on scroll.
Rows with a "details" attribute should expand to show a details pane.
Cells with a "wrap" attribute should allow for word wrapping.
Cells with a "disable" attribute should be disabled
Rows with a "disable" attribute should be disabled
Columns with a "disable" attribute should be disabled

### Relevant Info

Should be done after: #1574 

### Helpful Details

3rd party components currently in use at Esri include Vaadin Grid, Tabulator JS, MUI, and React DataGrid. If you are aware of other data grids in use please comment here or let the Calcite team know.

### Priority impact

p3 - want for upcoming milestone

### Esri team

Calcite (dev)

**Comment by forrestlin2020** [comment]
Where is Advanced DataGrid?

**Comment by geospatialem** [comment]
Design will be sought in the May milestone. 

Once design has been completed, will be re-routed for dev, targeted for the August release.

**Comment by yelenakreyndel** [comment]
@geospatialem advanced table will not be complete for the August release. The basic version is planned to be. We still don't have an estimate for the completion of the advanced version, and we need to gather it ASAP as advanced data grid has been a primary request for Calcite by Esri product teams for a long time. Design work for the advanced table will be complete in May as mentioned. Could you please get an estimate from Ali for the advanced version?

**Comment by geospatialem** [comment]
@yelenakreyndel Once design is complete in May, the milestone will be modified to reflect the upcoming development milestone. I reflected my comment above regarding the dev milestone.

**Comment by yelenakreyndel** [comment]
@geospatialem Your response is incorrect. There has been an in-depth planning for data-grid component. We have defined a milestone for implementation of a basic component already. There's no need to redefine that but we also need to gather an estimate for implementation of an advanced version which has no dependency on design. Are you able to help with getting us an estimate?

**Comment by geospatialem** [comment]
Research/design considerations will start in September, and development work will begin TBD in 2024.

**Comment by brittneytewks** [comment]
Removed figma changes label, will track progress for this work in monday.com

**Comment by LeviRemi** [comment]
Hello! I would like to request that the team considers a Multi-Column Sorting capability as part of the larger, advanced table functionality effort. I wanted to voice this early on, in case some of the design requirements may impact the early foundations of the "simple" sorting that is already planned.

The use-case is for larger tables where a user may want to sort by columns with duplicate values: e.g. sort by State, then City, then Street.

For UI/UX, I imagine "multiple-column-sorting" as an optional feature. You might consider sort-directions + enable-sort/disable-sort to be tied to the same icon in each column header. If multiple columns have been sort-activated, you could either assume the priority based on current column order, or you could render a number beside the icon to indicate what priority-order the sorting is occurring at. 

I think this implementation by Shield UI is a good example:
https://demos.shieldui.com/web/grid-general/sorting

**Comment by macandcheese** [comment]
Thanks @LeviRemi!

> I think this implementation by Shield UI is a good example: https://demos.shieldui.com/web/grid-general/sorting

One design consideration I can think of - which many other libraries don't do a great job of - is indicating which is the "first sort" vs "second sort" - MUI does an ... okay job, but it is effective enough. Maybe this is something we can improve on: 

<img width="893" alt="Screenshot 2023-12-21 at 2 45 49 PM" src="https://github.com/Esri/calcite-design-system/assets/4733155/2e2399ed-f857-4411-8a9a-2fd72cdc921d">

**Comment by LeviRemi** [comment]
@macandcheese - I share the same feelings. It's a difficult design to pull off. The only alternative I could think of is to relegate all column sorting actions to a Calcite Dropdown that contains a list with two groups: sorted/non-sorted. Users could click non-sorted fields to move them into the sorted group. The sorted group could be rearranged via drag-n-drop, and each sorted field could have action buttons to either toggle the sort direction or disable sort on that field - moving it to the non-sorted group.

To me, this alternative feels like it would be intuitive while the dropdown is open, but lacking context once collapsed. Just food for thought. Thanks for considering!

**Comment by alisonailea** [comment]
Another way to handle this would be to use natural language somewhere in the header.

Table A, displaying 250 results, sorted by ColA then ColB.

**Comment by benelan** [comment]
# Research summary

For the initial advanced table work, we'll start from scratch in a new component using [Tabulator](https://tabulator.info/docs/6.2/format#format-custom) to implement the rich table features. The existing `table-calcite` component is split into cell/row sub-components, which could complicate the integration of advanced functionality. The current goal is to make the advanced table's default behavior align with `calcite-table` so it can be a potential replacement.

## Requirements

Tabulator meets the [requirements](https://devtopia.esri.com/WebGIS/arcgis-web-components/discussions/1459#discussioncomment-7255) specified by the ArcGIS FeatureTable component developers. Note that not every point below will be fully implemented on our end. For example, we probably won't add builtin support for Arcade expressions or other ArcGIS specific functionality. However, will provide users with the means to implement out of scope features via access to Tabulator's customization options. A goal is to provide an example of how to implement any of the following requirements that aren't built into the component.

- Row virtualization [^1]
  - lazy loading of rows and associated data (500k+; no upper limit?)
- Column virtualization [^2]
  - Similar to rows; support for 100? 200+ columns
- Ability to select multiple rows (or turn off selection entirely) [^3]
  - (optional; can be custom) checkboxes in a special (pinned?) selection column
- Ability to sort columns [^4]
  - Sort by single OR multiple columns in ascending/descending order, in a given priority
  - e.g. sort by "last name" in ascending order then "first name" in ascending order
  - NOTE: typically up to the data source to sort data at request/render time
- Dynamic/custom cell content (determined at render time [^5a])
  - Examples:
    - Format raw values from the data source (e.g. unix date strings to "September 10th, 2024") [^5b]
    - Create components based on cell value or column type (e.g. calcite-dropdown, calcite-action, calcite-icon etc) [^5c]
    - HTML elements (divs and spans containing user managed content) [^5d]
    - Execute async arcade expressions on first render or get cached values to display [^5e]
  - Header and body cells
  - Footer cells not required
  - Note: sorting is still based on "raw" data values or disabled if applicable
- Ability to hide and show columns after creation (can be via custom menu) [^6]
- Custom column widths [^7a] and resizing columns via mouse/pointer in the UI [^7b]
- Reorder columns via drag-and-drop in the UI [^8]
- Support for nested/group columns [^9]
  - Group columns do not need to be sortable
- Ability to scroll [^10a] to a specific row by index [^10b]
- Ability to temporarily filter the table by specific objectIds (can be tied to data store) [^11]
- Ability to conditionally style individual cells or rows [^12]
  - e.g. background color of "highlighted" columns, or specific rows
- Mouse/keyboard interaction events [^13]
  - e.g. click, dbl-click, pointer-over, pointer-out, keydown
  - Ability to know which cell/column/row is being targeted
- Ability to "freeze" or "pin" columns to the left or right edge of the table (always visible) [^14]
- Editing support for cell content [^15a] [^15b]
  - Not immediately required as long as "Dynamic/custom cell content" and "cell events" are supported

[^1]: <https://tabulator.info/docs/6.2/virtual-dom>

[^2]: <https://tabulator.info/docs/6.3/layout#virtual-dom-hoz>

[^3]: <https://tabulator.info/docs/6.2/select>

[^4]: <https://tabulator.info/docs/6.2/sort>

[^5a]: <https://tabulator.info/docs/6.2/reactivity#reactive-data>

[^5b]: <https://tabulator.info/docs/6.2/format#formatter-datetime>

[^5c]: <https://tabulator.info/docs/6.2/format#format-custom>

[^5d]: <https://tabulator.info/docs/6.2/format#formatter-html>

[^5e]: <https://tabulator.info/docs/6.2/mutators#mutators>

[^6]: <https://tabulator.info/docs/6.2/columns#visibility>

[^7a]: <https://tabulator.info/docs/6.2/layout#layout>

[^7b]: <https://tabulator.info/docs/6.2/layout#resize-column>

[^8]: <https://tabulator.info/docs/6.2/move#columns>

[^9]: <https://tabulator.info/docs/6.2/columns#groups>

[^10a]: <https://tabulator.info/docs/6.2/navigation#scroll-row>

[^10b]: <https://tabulator.info/docs/6.2/data#row-index>

[^11]: <https://tabulator.info/docs/6.2/filter#search-data>

[^12]: <https://tabulator.info/docs/6.2/style#css-rows>

[^13]: <https://tabulator.info/docs/6.3/events>

[^14]: <https://tabulator.info/docs/6.2/layout#frozen-column>

[^15a]: <https://tabulator.info/docs/6.2/edit>

[^15b]: <https://tabulator.info/docs/6.2/modules#module-edit>

**Comment by driskull** [comment]
Since it seems like the advanced tables include pagination, I wonder if we would use our own pagination component in place of them or align our pagination to component to work and feel like whatever table component we use.

Some of the other things like menus, popovers, alerts we just would use our own components for.

**Comment by driskull** [comment]
@benelan do you know if its possible to sort or resize columns using a keyboard? Wondering what the accessibility is for some of the functionality.

**Comment by benelan** [comment]
It is possible for sort, here is a [demo](https://tabulator.info/examples/6.3#sorters). I'm not sure about resizing columns, it might require dragging with the mouse.

**Comment by benelan** [comment]
> Since it seems like the advanced tables include pagination, I wonder if we would use our own pagination component in place of them or align our pagination to component to work and feel like whatever table component we use.
> 
> Some of the other things like menus, popovers, alerts we just would use our own components for.

Yeah I think it depends on how different the behaviors are, but it shouldn't be too hard to use `calcite-pagination` instead.

**Comment by aPreciado88** [comment]
State of advanced table update.

We've made progress on the following requirements:

- Ability to sort columns
- Custom column widths and resizing columns via mouse/pointer in the UI
- Reorder columns via drag-and-drop in the UI
- Support for nested/group columns
- Ability to scroll to a specific row by index
- Ability to temporarily filter the table by specific object Ids
- Ability to conditionally style individual cells or rows
- Ability to "freeze" or "pin" columns to the left or right edge of the table (always visible)
- Load data from HTML table (simple table)

We are continuing to work on the remaining requirements and will provide more updates here when progress is made.

**Comment by forrestlin2020** [comment]
Hello Kitty,

I like the sessions of Calcite Design System: What's New and the Road Ahead, ArcGIS Maps SDK for JavaScript: App Development with Components, Part 3: User Experience and ArcGIS Maps SDK for JavaScript: App Development with Components, Part 4: Extending and Styling.

Is there any way I can get a copy of the slides?

Regards,

Forrest

Feng Forrest Lin
GIS Lead – Applications Developer
Representing Amy Mercado | Orange County Property Appraiser
200 S. Orange Ave |  Suite 1700 | Orlando, FL 32801
407.836.5281 work  | 407.605.0391 fax
***@***.******@***.***>   | www.ocpafl.org<http://www.ocpafl.org>

  [cid:ocpalogo_a7aa3e8b-8844-4655-a76f-e6588522815a.png] <https://www.ocpafl.org/>

[cid:asset2_2b5a9104-a103-4202-ab87-ec45545f574e.png]<https://www.ocpafl.org/>


 [cid:facebook_4fc6d226-27e2-4277-9c11-6f34177bbc26.png] <https://www.facebook.com/AmyOCPA>   [cid:x_787a21e3-caf5-4132-8e57-a77cc1bb3421.png] <https://x.com/AmyOCPA>   [cid:youtube_fefff1e6-f86c-43e9-a618-5ed9c05a8c99.png] <https://www.youtube.com/AmyOCPA>   [cid:instagram_4576db4c-244c-490b-89e4-8214f58265eb.png] <https://www.instagram.com/AmyOCPA/#>   ***@***.***_c04279a7-0e02-4b67-a60e-b907a33de340.png] <https://www.linkedin.com/company/AmyOCPA>   [cid:reddit_cb986e90-395f-40d4-b0c2-170a6e868cdb.png] <https://www.reddit.com/user/AmyOCPA/>    [cid:threads_4d598325-7160-4ce0-b8e7-d91b5ca3be98.png] ***@***.***>   [cid:bluesky_cef1cf06-d0ac-488a-9a96-4c6545c56883.png] <https://bsky.app/profile/amyocpa.bsky.social>
             [cid:asset17_7fe3ce62-0626-48df-a03a-e2696b151918.png]      [cid:asset18_b9dba45c-9a56-4093-8d49-7c8d82e47e3b.png]


@AmyOCPA

________________________________
This communication is intended only for the recipient(s) identified in the message.  Review, dissemination, or copying of this communication by someone other than the intended recipient(s) is prohibited by law.  If this communication was received in error, then please notify the sender and delete all copies of this communication.  Any tampering with or altering the contents of this message is prohibited by law.  This communication is treated the same as any written document and may be subject to all rules and laws governing public information and documents, including, without limitation, Article 1, Section 24, of the Florida State Constitution and Chapter 119, Florida Statutes.

________________________________

**Comment by geospatialem** [comment]
> Hello Kitty, I like the sessions of Calcite Design System: What's New and the Road Ahead, ArcGIS Maps SDK for JavaScript: App Development with Components, Part 3: User Experience and ArcGIS Maps SDK for JavaScript: App Development with Components, Part 4: Extending and Styling. Is there any way I can get a copy of the slides?

@forrestlin2020 The above issue relates to the advanced table new component, however followed up with the resources you are seeking on [this Community blog post](https://community.esri.com/t5/calcite-design-system-blog/2026-developer-amp-technology-summit-calcite/ba-p/1690703). Thanks for attending!

**Comment by macandcheese** [comment]
cc @ryhoran @Esri/calcite-pes - Sdk team is looking to turn their Feature Table into a component offering, this would be a good opportunity to sync and see if a shared, generic, not-sdk-specific advanced table could come out of that work.

**Comment by ameliaatesri** [comment]
The ArcGIS Monitor team would love to utilize the kind of advanced table described in this issue. The following items from your description are of particular interest:

- virtualize creating a Table from a large data set
- be able to easily render any component inside a Table component
- have a fast and performant way to find data in existing loaded data set
- be able to select table rows
- have lazy-loading on scroll of additional data rows
- have lazy-loading on scroll of additional data columns
- hide columns 
- resize or adjust the width of columns

In addition, we'd also want an advanced table component to support:
- column header actions 
- sticky headers + columns

Just wanted to leave a note of interest/support here, so we can follow progress on the issue 😄
